### PR TITLE
New feature/ccp 247

### PR DESCRIPTION
**Pull request by abhis1989kumar**

### CODE REVIEW

**Comment by jdamerow** [comment]
there seem to be a bunch of review request classes in this branch that do not belong to this ticket. please remove.

File: Conceptpower+Spring/pom.xml [hunk 1]
```diff
@@ -643,7 +612,7 @@
 				<version>2.2</version>
 				<configuration>
 					<url>${tomcat.path}</url>
-					<server>${tomcat.server.id}</server>
+					<server>TomcatServer</server>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml]: this needs to be configurable, hence the variable

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/jaxb/viaf/Channel.java [hunk 2]
```diff
@@ -14,8 +14,8 @@
  */
 @XmlAccessorType(XmlAccessType.FIELD)
 public class Channel {
-
-	@XmlElement(name = "item")
+    
+    @XmlElement(name = "item")
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/jaxb/viaf/Channel.java:18]: indentation

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/profile/impl/ViafService.java [hunk 3]
```diff
@@ -98,17 +109,33 @@ public String getName() {
...
+		ResponseEntity<ViafReply> reply;
+        try {
+            RestTemplate restTemplate = new RestTemplate();
+            reply = restTemplate.exchange(new URI(fullUrl),HttpMethod.GET, entity, ViafReply.class);
+        } catch (RestClientException | URISyntaxException e) {
+            logger.error("Error during contacting VIAF.", e);
+            return searchResults;
+        }
+		
+		ViafReply rep = reply.getBody();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/profile/impl/ViafService.java:136]: formatting

File: Conceptpower+Spring/src/main/resources/log4j.xml [hunk 4]
```diff
@@ -31,6 +31,10 @@
 	<logger name="org.springframework.web">
 		<level value="info" />
 	</logger>
+	
+	<logger name="org.springframework.security">
+		<level value="debug" />
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/resources/log4j.xml]: let's set that to info

Comment (jdamerow) [reply Conceptpower+Spring/src/main/resources/log4j.xml]: actually, let's use ${debug_level} for all debugger

File: Conceptpower+Spring/src/main/webapp/WEB-INF/spring/db4o-config.xml [hunk 5]
```diff
@@ -20,5 +20,9 @@
 	<bean id="luceneDatabaseManager" class="edu.asu.conceptpower.app.db.DatabaseManager" init-method="init" destroy-method="close">
 		<property name="databasePath" value="${db.path}/lucene.db" />
 	</bean>
+	
+  	<bean id="conceptReviewDatabaseManager" class="edu.asu.conceptpower.app.db.DatabaseManager" init-method="init" destroy-method="close">
+		<property name="databasePath" value="${db.path}/conceptReview.db" />
+	</bean>
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/spring/db4o-config.xml]: does not belong into this pull request

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/home.jsp [hunk 6]
```diff
@@ -5,23 +5,22 @@
 <%@taglib uri="http://www.springframework.org/tags/form" prefix="form"%>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/functions" prefix="fn"%>
 <%@ page session="false"%>
-
+<meta name="_csrf" content="${_csrf.token}"/>
+	<!-- default header name is X-CSRF-TOKEN -->
+	<meta name="_csrf_header" content="${_csrf.headerName}"/>
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/home.jsp]: indentation

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/nav/showresults.jsp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            }
+        }
+    });
+});
+
+	</script>
+
+<p> ${comment} <p>
+</body>
+</html>
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/nav/showresults.jsp]: what is this page for?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java [hunk 8]
```diff
@@ -80,10 +80,10 @@ public void testSearch() {
...
         List<ISearchResult> searchResults = viafService.search("Pirckheimer");
 
-        assertEquals(1, searchResults.size());
-        assertEquals("Mon, 12 Jul 2015 18:51:56 GMT", searchResults.get(0).getDescription());
-        assertEquals("http://viaf.org/viaf/27173507", searchResults.get(0).getId());
-        assertEquals("Pirckheimer, Willibald, 1470-1530.", searchResults.get(0).getName());
+        assertEquals(10, searchResults.size());
+        assertEquals("Mon, 18 Feb 2018 10:37:06 GMT", searchResults.get(0).getDescription());
+        assertEquals("http://viaf.org/viaf/4473151898622624190001", searchResults.get(0).getId());
+        assertEquals("Pirckheimer TIB X Kommentar.1001.106 C6", searchResults.get(0).getName());
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java]: how can set be? the search result set up on top returns something else?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Conceptpower+Spring/pom.xml [hunk 9]
```diff
@@ -643,7 +614,7 @@
 				<version>2.2</version>
 				<configuration>
 					<url>${tomcat.path}</url>
-					<server>${tomcat.server.id}</server>
+					<server>${tomcat.server}</server>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml]: that's not the right property

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/jaxb/viaf/Channel.java (hunk 2 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/jaxb/viaf/Channel.java:18]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]
please address all my previous comments before I finish reviewing

File: Conceptpower+Spring/pom.xml [hunk 10]
```diff
@@ -53,6 +54,7 @@
 				<db.path>${db.files.path}</db.path>
 				<deploy.path>${deploy.path.dev}</deploy.path>
 				<tomcat.path>${tomcat.deploy.path}</tomcat.path>
+				<tomcat.server>localhost</tomcat.server>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml]: Oh, I didn't realize you've added those. In that case please delete them.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/profile/impl/ViafService.java [hunk 11]
```diff
@@ -98,17 +113,34 @@ public String getName() {
...
 		List<Item> items = null;
-		String fullUrl;
-
-		fullUrl = viafURL.trim() + searchViafURLPath.trim() + " " + word.trim()
-				+ searchViafURLPath1.trim() + startIndex.trim()
-				+ searchViafURLPath2.trim();
-
-		ViafReply rep = (ViafReply) restTemplate.getForObject(fullUrl,
-				ViafReply.class);
+		String fullUrl="";
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/app/profile/impl/ViafService.java]: declare when needed

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java [hunk 12]
```diff
@@ -24,23 +36,26 @@
...
+    private RestTemplate template;
 
+    @Mock
+    private ISearchResultFactory searchResultFactory;
+    
     @InjectMocks
     private ViafService viafService;
+    
+    @Mock
+    private RestTemplate restTemplate;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java:48]: there are two restTemplates now, but I would think you only need one

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java [hunk 13]
```diff
@@ -70,9 +96,16 @@ public void setUp() {
...
+                "http://viaf.org/viaf/search?query=local.names%20all%20Test+for+Null+&amp;maximumRecords=100&amp;startRecord=1&amp;sortKeys=holdingscount&amp;httpAccept=application/rss+xml",
                 ViafReply.class)).thenReturn(emptyViafReply);
 
+        
+        reply = new ResponseEntity<ViafReply>(emptyViafReply, HttpStatus.OK);
+        try {
+            Mockito.when(restTemplate.exchange(new URI("http://viaf.org/viaf/search?query=local.names%20all%20Test+for+Null+&amp;maximumRecords=100&amp;startRecord=1&amp;sortKeys=holdingscount&amp;httpAccept=application/rss+xml"),HttpMethod.GET, entity, ViafReply.class)).thenReturn(reply);
+        } catch (RestClientException | URISyntaxException e) {
+            logger.error("Error during contacting VIAF.", e);
+        }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java:105]: duplicate code (line 80-89)?

File: Conceptpower+Spring/src/test/resources/log4j.xml [hunk 14]
```diff
@@ -31,6 +31,9 @@
 	<logger name="org.springframework.web">
 		<level value="info" />
 	</logger>
+	<logger name="org.springframework.security">
+		<level value="debug" />
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/resources/log4j.xml]: let's use the pom property here, we don't want it to always debug

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java (hunk 12 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java:48]: ?

**Comment by jdamerow** [comment]
my previous comments don't seem to be addressed.

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java [hunk 15]
```diff
@@ -24,23 +36,23 @@
...
+    
+    @Mock
+    private RestTemplate restTemplate;
 
     @Before
     public void setUp() {
-        template = Mockito.mock(RestTemplate.class);
-
-        MockitoAnnotations.initMocks(this);
+          MockitoAnnotations.initMocks(this);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java]: indentation?

File: Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java (hunk 13 above)
Comment (jdamerow) [reply Conceptpower+Spring/src/test/java/edu/asu/conceptpower/profile/impl/ViafServiceTest.java:105]: ?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

File: Conceptpower+Spring/pom.xml [hunk 16]
```diff
@@ -292,136 +324,101 @@
...
-			<version>4.2.0.Final</version>
+			<version>6.0.15.Final</version>
 		</dependency>
 
 		<!-- DB4O -->
 		<dependency>
 			<groupId>com.db4o</groupId>
 			<artifactId>db4o-full-java5</artifactId>
-			<version>8.0</version>
+			<version>8.0.276.16149</version>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml]: please do not change that version

**Comment by jdamerow** [comment]
integration tests are failing, please fix

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

**Comment by jdamerow** [comment]
I don't know why the integration tests pass locally but not on Jenkins, but the issue is that something goes wrong with the indexing. The tests that fail, fail because of this exception:
```
ERROR: edu.asu.conceptpower.web.util.ExceptionHandler - ExceptionHandler caught exception.
java.lang.IllegalStateException: field "listName" was indexed without position data; cannot run PhraseQuery (phrase=listName:"vogonweb concepts")
	at org.apache.lucene.search.PhraseQuery$PhraseWeight.scorer(PhraseQuery.java:430)
	at org.apache.lucene.search.BooleanWeight.scorer(BooleanWeight.java:280)
	at org.apache.lucene.search.Weight.bulkScorer(Weight.java:135)
	at org.apache.lucene.search.BooleanWeight.bulkScorer(BooleanWeight.java:262)
```
If you add ```.param("concept_list", "VogonWeb Concepts")``` to the test case ```test_searchConcept_searchWithWordAndPosInJson``` you will get the same exception locally (for the searchWithWordAndPosInJson test case, the noResults test cases still pass for some reason):
```
@Test
    public void test_searchConcept_searchWithWordAndPosInJson() throws Exception {

        final String output = IOUtils
                .toString(this.getClass().getClassLoader().getResourceAsStream("output/conceptWithWordAndPos.json"), "UTF-8");

        this.mockMvc
                .perform(MockMvcRequestBuilders.get("/ConceptSearch").param("word", "einstein").param("pos", "noun").param("concept_list", "VogonWeb Concepts")
                        .param("operator", SearchParamters.OP_AND).accept(MediaType.APPLICATION_JSON_VALUE))
                .andExpect(content().json(output, false)).andExpect(status().isOk());

    }
```

**Comment by jdamerow** [comment]
test this please

**Comment by jdamerow** [comment]
Jenkins, run integration tests please

**Comment by jdamerow** [comment]
test this please

File: Conceptpower+Spring/pom.xml [hunk 17]
```diff
@@ -9,15 +10,15 @@
...
 		<java-version>1.8</java-version>
-		<org.springframework-version>4.3.18.RELEASE</org.springframework-version>
-		<org.aspectj-version>1.7.3</org.aspectj-version>
-		<org.slf4j-version>1.7.5</org.slf4j-version>
+		<org.springframework-version>5.2.0.RELEASE</org.springframework-version>
+		<org.aspectj-version>1.9.2</org.aspectj-version>
+		<org.slf4j-version>1.7.26</org.slf4j-version>
+		<spring-security-version>5.2.0.RELEASE</spring-security-version>
 		<tomcat.deploy.path>http://localhost:8080/manager/text</tomcat.deploy.path>
-		<tomcat.server.id>TomcatServer</tomcat.server.id>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml:16]: should not be deleted as it is breaking the build

**Comment by jdamerow** [comment]
test this please

File: Conceptpower+Spring/pom.xml [hunk 18]
```diff
@@ -641,11 +611,6 @@
 				<groupId>org.apache.tomcat.maven</groupId>
 				<artifactId>tomcat7-maven-plugin</artifactId>
 				<version>2.2</version>
-				<configuration>
-					<url>${tomcat.path}</url>
-					<server>${tomcat.server.id}</server>
-					<path>/${deploy.path}</path>
-				</configuration>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml:648]: can't be removed

Comment (kkrish39) [reply Conceptpower+Spring/pom.xml:648]: Added it back

**Pull request merged**